src/cargo/util/graph.rs

```diff
@@ -154,17 +333,271 @@ impl<N: fmt::Display + Eq + Ord + Clone, E: Clone> fmt::Debug for Graph<N, E> {
+        N: Borrow<Q>,
+        Q: Hash + Eq,
+    {
+        self.borrow()
+            .inner
+            .nodes
+            .get_full(k)
+            .filter(|(i, _, _)| *i < self.age.len_nodes)
+            .is_some()
+    }
+
+    /// Checks if there is a path from `from` to `to`.
+    pub fn is_path_from_to<'a>(&'a self, from: &'a N, to: &'a N) -> bool {
+        let mut stack = vec![from];
+        let mut seen = HashSet::new();
+        let inner = &self.borrow().inner;
+        seen.insert(from);
+        while let Some((_, _, iter)) = stack.pop().and_then(|p| {
+            inner
+                .nodes
+                .get_full(p)
+                .filter(|(i, _, _)| *i < self.age.len_nodes)
+        }) {
+            for (p, idx) in iter.iter() {
+                if *idx < self.age.len_edges {
+                    if p == to {
+                        return true;
+                    }
+                    if seen.insert(p) {
+                        stack.push(p);
+                    }
+                }
+            }
+        }
+        false
+    }
+}
+
+impl<N: Eq + Hash + Clone, E: Clone + PartialEq> Default for StackGraph<N, E> {
+    fn default() -> StackGraph<N, E> {
+        StackGraph::new()
+    }
+}
+
+impl<N: fmt::Display + Eq + Hash + Clone, E: Clone + fmt::Debug + PartialEq> fmt::Debug
+    for StackGraph<N, E>
+{
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        writeln!(fmt, "Graph {{")?;
+
+        for (from, e) in self.borrow().inner.nodes.iter().take(self.age.len_nodes) {
+            writeln!(fmt, "  - {}", from)?;
+
+            for (to, idx) in e.iter() {
+                if *idx < self.age.len_edges {
+                    writeln!(fmt, "    - {}", to)?;
+                    for edge in self.borrow().edge(from, to) {
+                        writeln!(fmt, "      - {:?}", edge)?;
+                    }
+                }
+            }
+        }
+
+        write!(fmt, "}}")?;
+
+        Ok(())
+    }
+}
+
+/// And Iterator of the edges related to one link in a `Graph` or a `StackGraphView`.
+#[derive(Clone, Debug)]
+pub struct Edges<'a, E: Clone> {
+    /// The arena the linked list is stored in. If we are used with a `StackGraphView`
+    /// then this is only the related prefix of the arena. So we need to filter out
+    /// pointers past the part we are given.
+    graph: &'a [EdgeLink<E>],
+    index: Option<EdgeIndex>,
+}
+
+impl<'a, E: Clone> Edges<'a, E> {
+    pub fn is_empty(&self) -> bool {
+        self.index
+            .and_then(|idx| self.graph.get(idx))
+            .and_then(|l| l.value.as_ref())
+            .is_none()
+    }
+}
+
+impl<'a, E: Clone> Iterator for Edges<'a, E> {
+    type Item = &'a E;
+
+    fn next(&mut self) -> Option<&'a E> {
+        while let Some(edge_link) = self.index.and_then(|idx| {
+            // Check that the `idx` points to something in `self.graph`. It may not if we are
+            // looking at a smaller prefix of a larger graph.
```

Comment:
I think this is something I don't quite understand, I thought the point of `reset_to` was that these sort of extraneous edges were pruned out? How do they linger around?